### PR TITLE
MUMMNG-2017: Make list-of-links widget type, LOL

### PR DESCRIPTION
**Pull request by jhanstra**
Adds a list-of-links widget type to /web. This widget just displays a grid of four links using the circle link directive. The information about these links (href, title, target, icon) are provided by the `widgetConfig`. Entity file changes for other list-of-link widgets coming soon.
image

### CODE REVIEW

File: angularjs-portal-home/src/main/webapp/my-app/layout/widget/controllers.js [hunk 1]
```diff
@@ -4,6 +4,9 @@ define(['angular'], function(angular){
 
   var app = angular.module('my-app.layout.widget.controllers', []);
 
+  app.controller('ListOfLinksController', ['$scope', 'layoutService', function($scope, layoutService){
+  }]);
```
Comment (timlevett) [inline angularjs-portal-home/src/main/webapp/my-app/layout/widget/controllers.js]: if there is nothing to do in the controller, you don't need one true? (directives don't have to have a controller connected to them)

File: angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+         <p>Docs</p>
+      </div>
+      <div class='col-xs-6 center'>
+         <div class='btn btn-primary rounded icon-button text-right'> <a href='https://hangouts.google.com/' target="_blank"><i class='fa fa-comment'></i></a> </div>
+         <p>Hangouts</p>
+      </div>
+      <div class='col-xs-6 center'>
+         <div class='btn btn-primary rounded icon-button text-right'> <a href='https://plus.google.com/' target="_blank"><i class='fa fa-google-plus'></i></a> </div>
+         <p>Google+</p>
+      </div> -->
```
Comment (timlevett) [inline angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html]: Comments?

File: angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        <loading-gif data-object='portlet.widgetData'></loading-gif>
+    </div>
+    <div ng-if="config.singleElement && (!portlet.widgetData[config.arrayName] || portlet.widgetData[config.arrayName].length == 0)" class='center'>
+        <span title='{{portlet.widgetData[config.display].length > 26 ? portlet.widgetData[config.display] : ""}}'>{{portlet.widgetData[config.display] | trimMiddle:26}}</span>
+    </div>
+    <div ng-if="portlet.widgetData && portlet.widgetData[config.arrayName] && portlet.widgetData[config.arrayName].length >= 1" style='margin-top: -10px;'>
+       
+    </div>
+  </div>
+</div> -->
```
Comment (timlevett) [inline angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html]: Remove?

File: angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html [hunk 4]
```diff
@@ -0,0 +1,35 @@
+<div class='widget-body'>
+   <div class='row'>
+      <div class='col-xs-6 center' ng-repeat='item in config.links'>
+				<circle-button data-href='{{item.href}}' data-target='{{item.target}}' data-fa-icon='{{item.icon}}' data-disabled='false' data-title='{{item.title}}'></circle-button>
```
Comment (timlevett) [inline angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html]: tabs vs spaces?

File: angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html [hunk 5]
```diff
@@ -0,0 +1,35 @@
+<div class='widget-body'>
+   <div class='row'>
+      <div class='col-xs-6 center' ng-repeat='item in config.links'>
```
Comment (timlevett) [inline angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html]: limit 4? 

What should 3 look like? Triangle? What should 2 look like, centered? One?

File: angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+         <div class='btn btn-primary rounded icon-button text-right'> <a href='https://hangouts.google.com/' target="_blank"><i class='fa fa-comment'></i></a> </div>
+         <p>Hangouts</p>
+      </div>
+      <div class='col-xs-6 center'>
+         <div class='btn btn-primary rounded icon-button text-right'> <a href='https://plus.google.com/' target="_blank"><i class='fa fa-google-plus'></i></a> </div>
+         <p>Google+</p>
+      </div> -->
+   </div>
+</div>
+<a class="btn btn-default launch-app-button" href="{{config.launchUrl}}" target="_blank">{{config.launchText}}</a>
```
Comment (timlevett) [inline angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html]: Reason we are not using the `portlet.url`?

File: angularjs-portal-home/src/main/webapp/my-app/layout/widget/controllers.js (hunk 1 above)
Comment (jhanstra) [inline angularjs-portal-home/src/main/webapp/my-app/layout/widget/controllers.js]: Right. Kind of put it in there as a placeholder but we could definitely take it out til it's necessary.

File: angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html (hunk 5 above)
Comment (jhanstra) [inline angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html]: Hmm, didn't think about that. I'll refactor to account for different numbers of links.

File: angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html (hunk 2 above)
Comment (jhanstra) [inline angularjs-portal-home/src/main/webapp/my-app/layout/widget/partials/lol.html]: whoops, forgot to get rid of those

File: angularjs-portal-home/src/main/webapp/css/widget.less [hunk 7]
```diff
@@ -80,4 +80,8 @@ weather {
 .scroll-widget {
   max-height: 210px; 
   overflow-y: scroll;
+}
+.lol .v-center {
```
Comment (timlevett) [inline angularjs-portal-home/src/main/webapp/css/widget.less]: :100: for namespacing

**Comment by jhanstra** [comment]
Special cases: 
image

image

image

**Comment by timlevett** [comment]
This is huge! :+1: Strong work.

**Comment by jhanstra** [comment]
One more special case (more than 4 links):
image

**Comment by apetro** [comment]
:+1:  Hurray for templates.  *This* is who we will scale and make maintenance affordable.

**Comment by keirserrie** [comment]
:+1: looks great, plus 7 means no scrolling woo hoo

**Pull request merged**